cpp/include/cudf/ast/expressions.hpp

```diff
@@ -552,6 +554,98 @@ class column_name_reference : public expression {
   std::string column_name;
 };
 
+/**
+ * @brief An AST expression tree. It owns and contains multiple dependent expressions. All the
+ * expressions are destroyed when the tree is destructed.
+ */
+class tree {
+ public:
+  /**
+   * @brief construct an empty ast tree
+   */
+  tree() = default;
+
+  /**
+   * @brief Moves the ast tree
+   */
+  tree(tree&&) = default;
+
+  /**
+   * @brief move-assigns the AST tree
+   * @returns a reference to the move-assigned tree
+   */
+  tree& operator=(tree&&) = default;
+
+  ~tree() = default;
+
+  // the tree is not copyable
+  tree(tree const&)            = delete;
+  tree& operator=(tree const&) = delete;
+
+  /**
+   * @brief Add an expression to the AST tree
+   * @param args Arguments to use to construct the ast expression
+   * @returns a reference to the added expression
+   */
+  template <typename Expr, typename... Args>
+  Expr const& emplace(Args&&... args)
+  {
+    static_assert(std::is_base_of_v<expression, Expr>);
+    auto expr            = std::make_shared<Expr>(std::forward<Args>(args)...);
+    Expr const& expr_ref = *expr;
+    expressions.emplace_back(std::static_pointer_cast<expression>(std::move(expr)));
+    return expr_ref;
+  }
+
+  /**
+   * @brief Add an expression to the AST tree
+   * @param expr AST expression to be added
+   * @returns a reference to the added expression
+   */
+  template <typename Expr>
+  Expr const& push(Expr expr)
+  {
+    return emplace<Expr>(std::move(expr));
+  }
+
+  /**
+   * @brief get the first expression in the tree
+   * @returns the first inserted expression into the tree
+   */
+  expression const& front() const { return *expressions.front(); }
+
+  /**
+   * @brief get the last expression in the tree
+   * @returns the last inserted expression into the tree
+   */
+  expression const& back() const { return *expressions.back(); }
+
+  /**
+   * @brief get the number of expressions added to the tree
+   * @returns the number of expressions added to the tree
+   */
+  size_t size() const { return expressions.size(); }
+
+  /**
+   * @brief get the expression at an index in the tree. Index is checked.
+   * @param index index of expression in the ast tree
+   * @returns the expression at the specified index
+   */
+  expression const& at(size_t index) { return *expressions.at(index); }
+
+  /**
+   * @brief get the expression at an index in the tree. Index is unchecked.
+   * @param index index of expression in the ast tree
+   * @returns the expression at the specified index
+   */
+  expression const& operator[](size_t index) const { return *expressions[index]; }
+
+ private:
+  // TODO: use better ownership semantics, the shared_ptr here is redundant. Consider using a bump
+  // allocator with type-erased deleters.
+  std::vector<std::shared_ptr<expression>> expressions;
```

Comment:
Why not a vector of `unique_ptr`? 
is `shared_ptr` used for future shared ownership?

Reply:
I see this was marked resolved, but I'm not sure why.

I think maybe `shared_ptr` makes it easier to reuse common expressions in the same tree, or perhaps reuse the expressions across operator trees? Normally `unique_ptr` would be our preference but I could see use for copy-ability of the expressions.

Reply:
also discussed above: https://github.com/rapidsai/cudf/pull/17156#discussion_r1825143679

Reply:
Thanks. "unique_ptr can't work with type-erased objects" is what I was looking for.